core/engine/src/builtins/date/utils.rs

```diff
@@ -805,20 +819,52 @@ impl<'a> DateParser<'a> {
         }
     }
 
-    fn next_expect(&mut self, expect: char) -> Option<()> {
+    fn next_expect(&mut self, expect: u8) -> Option<()> {
+        self.input
+            .next()
+            .and_then(|c| if *c == expect { Some(()) } else { None })
+    }
+
+    fn next_ascii_digit(&mut self) -> Option<u8> {
         self.input
             .next()
-            .and_then(|c| if c == expect { Some(()) } else { None })
+            .and_then(|c| if c.is_ascii_digit() { Some(*c) } else { None })
+    }
+
+    fn next_n_ascii_digits<const N: usize>(&mut self) -> Option<[u8; N]> {
+        let mut digits = [0; N];
+        for digit in &mut digits {
+            *digit = self.next_ascii_digit()?;
+        }
+        Some(digits)
     }
 
-    fn next_digit(&mut self) -> Option<u8> {
-        self.input.next().and_then(|c| {
-            if c.is_ascii_digit() {
-                Some((u32::from(c) - u32::from('0')) as u8)
-            } else {
-                None
+    #[allow(clippy::items_after_statements)]
+    #[allow(clippy::inline_always)]
+    fn parse_n_ascii_digits<const N: usize>(&mut self) -> Option<u64> {
```

Comment:
suggestion: add some tests for this function.

The optimizations here are kinda gnarly lol It would be nice to have some localized Rust tests just to be able to easily verify the behavior and correctness whenever someone comes back to it.